climada/util/lines_polys_handler.py

```diff
@@ -0,0 +1,811 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU Lesser General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+"""
+import logging
+import copy
+from collections import Counter
+
+import cartopy.crs as ccrs
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import pyproj
+import scipy as sp
+import shapely as sh
+import shapely.geometry as shgeom
+
+from climada.engine import Impact
+from climada.util import coordinates as u_coord
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def calc_geom_impact(
+        exp, impf_set, haz, res,
+        to_meters=False, disagg_met='avg', disagg_val=None, agg='sum'):
+    """
+    Compute impact for exposure with (multi-)polygons and/or (multi-)lines.
+    Lat/Lon values in exp.gdf are ignored, only exp.gdf.geometry is considered.
+
+    The geometries are first disaggregated to points. Polygons: grid with
+    resolution res*res. Lines: points along the line separated by distance res.
+    The impact per point is then re-aggregated for each geometry.
+
+    Parameters
+    ----------
+    exp : Exposures
+        The exposure instance with exp.gdf.geometry containing (multi-)polygons
+        and/or (multi-)lines
+    impf_set : ImpactFuncSet
+        The set of impact functions.
+    haz : Hazard
+        The hazard instance.
+    res : float
+        Resolution of the disaggregation grid (polygon) or line (lines).
+    to_meters : bool, optional
+       If True, res is interpreted as meters, and geometries are projected to
+       an equal area projection for disaggregation.  The exposures are
+       then projected back to the original projections before  impact
+       calculation. The default is False.
```

Comment:
I think I prefer the `to_meters` version because one can use any unit in principle. When `to_meters = False`, the disaggregation is simply done on the coordinates provided, and thus `res` is in the same unit as the provided coordinates. In this way, the user can use any projection for their coordinate. This needs neither be in `meters` nor in `degrees`.